### ISSUE TITLE
Maps projection - make coordinate system configurable

### ISSUE THREAD

**Issue by joha78** [sothawo/mapjfx#46]
Thank you very much for this great project. Do you have an idea whether it is possible to apply an WGS 84 (EPSG:4326) transformation/projection to OpenLayers? 

https://openlayers.org/en/master/examples/reprojection.html

In this example they are using a OSM map, I wonder if this would work with bing maps too (mainly because OSM has tile loading problems still)

**Comment by sothawo** [comment]
not sure I understand what you mean. At the moment MapView coordinates are in WGS84(EPSG:4326) and are in the Javascript converted to EPSG:3857 which are the default coordinates used by OpenLayers.

**Comment by joha78** [comment]
Oh, I didn't know this. As I understand this means that the map data is actually in WGS84(EPSG:4326) but the view is transformed into EPSG:3857 by OpenLayers Javascript? Can you give me a hint how to view the map in  WGS84(EPSG:4326) ?

**Comment by sothawo** [comment]
at the moment this conversion is hard coded in the Javascript part. One solution would be to make both coordinate system configurable and pass these configured values into the javascript layer. I added the _enhancement_ label to this issue and changed the title accordingly.

**Comment by joha78** [comment]
I would like contributing this. But honestly I don't know where to start. I searched the whole project for "EPSG:" and "projection" but could not find a place that made sense... 

The most relevant part I could find was in mapview.js. I added the row "projection: 'EPSG:4326'"  but than the map does not load anymore:

```
if (newType == 'BINGMAPS_ROAD') {
        _map.setLayerGroup(new ol.layer.Group({
            layers: [
                new ol.layer.Tile({
                    source: new ol.source.BingMaps({
                        imagerySet: 'Road',
                        key: this.bingMapsApiKey,
                        projection: 'EPSG:4326'
                    })
                }),
                _layerFeatures
            ]
        }));
    }
```

**Comment by sothawo** [comment]
there is a functions.js file where at the moment coordinates are changed from one system to the other, this would need to be adapted as well. Basically I would add a converter object in javascript which gets the two strings/formats for conversion and refactor the functions.js functions into that object. The values for this converter object should be passed in from the Java side.

If you want to try this out please branch from 1.x if you're on JDK8, 2.x otherwise. I don't think I will have time before Sunday to look into this

**Comment by sothawo** [comment]
I created a branch _feature/projections_ from the 1.x branch (JDK8 based). In this branch I refactored the JavaScript part to have the projections used in OpenLayers and for the MapView to be configurable. I added a method in the `MapView` class to set these values (at the moment I have just defined a couple of values in the `Projection` enum.
The values are set with

    setProjections(WGS_84, WEB_MERCATOR);

first value defining the values used in Java, the second the internal values for OpenLayers. The shown values are the defaults that were hardcoded up to now: EPSG:4326 for the coordinates in the `MapView` and EPSG:3857 used by OpenLayers.

Can you check out this branch and see if this fits your needs?

**Comment by sothawo** [comment]
The first value is the value used for the coordinate values in MapView; according to https://en.wikipedia.org/wiki/Web_Mercator_projection this should always be WGS84. The second value is the one that is passed as projection value to OpenLayers. I can see no difference there as well, but when I read about the difference between WGS84 and Web Mercator, is this really such a big difference? I'm not expert in mapping technologies, so I don't know how a given map extend would differ in these systems.
The examples in the OpenLayers docs do not only switch the projection but also the map style and extent, so it's hard to see the difference.

**Comment by joha78** [comment]
Oh dear, I am sorry. I wrote  setProjections( Projection.WGS_84, Projection.WGS_84) here, but it was meant to be setProjections(Projection.WGS_84, Projection.EPSG_4326). With Projection.EPSG_4326 there should be a difference. Specially Greenland should be compressed vertically with EPSG_4326 but isn't. Have a look at https://openlayers.org/en/latest/examples/epsg-4326.html

**Comment by sothawo** [comment]
I have managed to get a map in EPSG:4326, but only when setting this projection initially in the javascript code. I have not found a way to change it dynamically. And I still have to find a way to be able to specify this value in the Java class and passing it into the javascript part before the _map object is created there. So it seems the most I can achieve at the moment is a MapView with a different projection which is set initially and cannot be changed afterwards.
But not today.

**Comment by joha78** [comment]
Wow, that sounds pretty good to me. For my use case setting it initially would  be sufficient.

**Comment by sothawo** [comment]
I have it implemented in the branch and will merge it into a new release on Friday. No time before.

To use a different projection all you have to do is pass a corresponding argument to the initialize method:

 ` mapView.initialize(Projection.WGS_84);`

attached a screenshot from the demo program using WGS84 and a Bing map.

![bildschirmfoto 2018-12-04 um 22 50 49](https://user-images.githubusercontent.com/8456476/49475497-b20e0680-f817-11e8-84c6-79dde7f288ab.png)

**Comment by sothawo** [comment]
implemented in mapjfx 1.22.0 and 2.3.0 and in the corresponding versions of mapjfx-demo. I will update the documentation on my website on friday.
